Support: flow-typed install resolve error

When running the `flow-typed install -o` it will run into an error from the resovle module:
`UNCAUGHT ERROR: Error: Cannot find module 'babel-runtime' from '/path/to/root-project-dir'`

What might be the best way to debug this issue? Is it coming from the cli or my project?

Here's the exact version

```
$ flow-typed version
flow-typed@2.0.0-beta.9
```


Here's the stack trace I can get from there:

```
UNCAUGHT ERROR: Error: Cannot find module 'babel-runtime' from '/path/to/project'
    at /flow-typed/node_modules/resolve/lib/async.js:46:17
    at process (/flow-typed/node_modules/resolve/lib/async.js:173:43)
    at ondir (/flow-typed/node_modules/resolve/lib/async.js:188:17)
    at load (/flow-typed/node_modules/resolve/lib/async.js:69:43)
    at onex (/flow-typed/node_modules/resolve/lib/async.js:92:31)
    at /flow-typed/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
```

I get the same error, but for react-scripts. It looks like it's not an install problem with the CLI, but a problem with what it's doing with my actual project dependencies?

```
wolfpuppy ~/projects/foo$ flow-typed install
• Found flow-bin@v0.32.0 installed. Installing libdefs compatible with this version of Flow...
• Found 9 dependencies in package.json. Searching for libdefs...
• Installing 2 libdefs...
  • flow-typed/npm/react-redux_v4.x.x.js already exists!
    └> Use --overwrite to overwrite the existing libdef.
  • flow-typed/npm/redux_v3.x.x.js already exists!
    └> Use --overwrite to overwrite the existing libdef.
UNCAUGHT ERROR: Error: Cannot find module 'react-scripts' from '/home/torne/projects/foo'
    at /home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:46:17
    at process (/home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:173:43)
    at ondir (/home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:188:17)
    at load (/home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:69:43)
    at onex (/home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:92:31)
    at /home/torne/.npm-global/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
```

Getting the same problem with Grunt deps:

```
• Found flow-bin@v0.33.0 installed. Installing libdefs compatible with this version of Flow...
• Found 34 dependencies in package.json. Searching for libdefs...
• rebasing flow-typed cache...done.
UNCAUGHT ERROR: Error: Cannot find module 'grunt-concurrent' from '/home/christopher/IdeaProjects/flow-angular'
    at /usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:46:17
    at process (/usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:92:31)
    at /usr/lib/node_modules/flow-typed/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
```

My package.json:

```
{
  "name": "yeomangeneratedproject",
  "private": true,
  "devDependencies": {
    "autoprefixer-core": "^5.2.1",
    "flow-bin": "^0.33.0",
    "grunt": "^0.4.5",
    "grunt-angular-templates": "^0.5.7",
    "grunt-concurrent": "^1.0.0",
    "grunt-contrib-clean": "^0.6.0",
    "grunt-contrib-compass": "^1.0.0",
    "grunt-contrib-concat": "^0.5.0",
    "grunt-contrib-connect": "^0.9.0",
    "grunt-contrib-copy": "^0.7.0",
    "grunt-contrib-cssmin": "^0.12.0",
    "grunt-contrib-htmlmin": "^0.4.0",
    "grunt-contrib-imagemin": "^1.0.0",
    "grunt-contrib-jshint": "^0.11.0",
    "grunt-contrib-uglify": "^0.7.0",
    "grunt-contrib-watch": "^0.6.1",
    "grunt-filerev": "^2.1.2",
    "grunt-google-cdn": "^0.4.3",
    "grunt-jscs": "^1.8.0",
    "grunt-newer": "^1.1.0",
    "grunt-ng-annotate": "^0.9.2",
    "grunt-postcss": "^0.5.5",
    "grunt-svgmin": "^2.0.0",
    "grunt-usemin": "^3.0.0",
    "grunt-wiredep": "^2.0.0",
    "jasmine-core": "^2.5.2",
    "jit-grunt": "^0.9.1",
    "jshint-stylish": "^1.0.0",
    "karma": "^1.3.0",
    "karma-jasmine": "^1.0.2",
    "karma-phantomjs-launcher": "^1.0.2",
    "phantomjs-prebuilt": "^2.1.12",
    "time-grunt": "^1.0.0"
  },
  "engines": {
    "node": ">=0.10.0"
  },
  "scripts": {
    "test": "karma start test/karma.conf.js"
  },
  "dependencies": {
    "lodash": "^4.16.2"
  }
}
```

--- a/cli/package.json
+++ b/cli/package.json
@@ -11,7 +11,7 @@
     "lint": "eslint .",
     "prepublish": "mkdir -p dist; npm run test",
     "test": "npm run clean; npm run build; npm run test-quick",
-    "test-quick": "npm run lint && jest",
+    "test-quick": "jest && npm run lint",
     "watch": "mkdir -p dist; babel --source-maps --watch=./src --out-dir=./dist"
   },
   "repository": {

--- a/cli/src/lib/npmProjectUtils.js
+++ b/cli/src/lib/npmProjectUtils.js
@@ -2,6 +2,7 @@
 
 import {fs} from "./node.js";
 import {path} from "./node.js";
+import resolveAsync from "resolve";
 import {searchUpDirPath} from "./fileUtils.js";
 import semver from "semver";
 import {stringToVersion} from "./semver.js";
@@ -119,4 +120,12 @@ export async function determineFlowVersion(
   }
 
   return null;
+};
+
+export function resolve(name: string, options: Object): Promise<string> {
+  return new Promise((res, rej) =>
+    resolveAsync(name, options, (err, result) => {
+      if (err) { rej(err); } else { res(result); }
+    })
+  );
 };

--- a/cli/src/lib/stubUtils.js
+++ b/cli/src/lib/stubUtils.js
@@ -10,7 +10,7 @@ import {getPackageJsonData} from './npmProjectUtils';
 import {getPackageJsonDependencies} from './npmProjectUtils';
 import {mkdirp} from './fileUtils';
 import {path} from './node';
-import resolveAsync from 'resolve';
+import {resolve} from "./npmProjectUtils";
 import {signCode} from './codeSign';
 import {verifySignedCode} from './codeSign';
 import {versionToString} from './semver';
@@ -27,16 +27,52 @@ export function glob(pattern: string, options: Object): Promise<Array<string>> {
   );
 }
 
-function resolve(name: string, options: Object): Promise<string> {
-  return new Promise((resolve, reject) =>
-    resolveAsync(name, options, (err, result) => {
-      if (err) {
-        reject(err);
-      } else {
-        resolve(result);
+async function resolveToDirPath(
+  name: string,
+  options: Object
+): Promise<string> {
+  let pkgFiles = [];
+  let opts = {
+    ...options,
+    packageFilter: function(pkgJson, pkgFilePath) {
+      if (!pkgJson.main) {
+        pkgFiles.push(path.join(path.dirname(pkgFilePath), 'index.js'));
+        pkgJson.main = "index.js";
       }
-    })
-  );
+      return pkgJson;
+    },
+    isFile: function(filePath, cb) {
+      if (pkgFiles.indexOf(filePath) !== -1) {
+        cb(null, true);
+      } else {
+        fs.stat(filePath)
+          .catch(e => {
+            if (e.code === 'ENOENT') {
+              cb(null, false);
+            } else {
+              cb(e);
+            }
+          })
+          .then(stat => {
+            cb(null, stat.isFile() || stat.isFIFO());
+          });
+      }
+    },
+  };
+  let resolvedPath = await resolve(name, opts);
+
+  let pathToPackage = path.dirname(resolvedPath);
+  let dir = pathToPackage;
+  while (path.basename(dir) !== "node_modules") {
+    pathToPackage = dir;
+    dir = path.dirname(pathToPackage);
+
+    if (dir === pathToPackage) {
+      throw new Error("Failed to find package directory");
+    }
+  }
+
+  return pathToPackage;
 }
 
 const moduleStubTemplate = (`
@@ -92,7 +128,7 @@ async function writeStub(
 
     const [fileDecls, aliases] = files.reduce(([fileDecls, aliases], file) => {
       const ext = path.extname(file);
-      const name = path.basename(file, ext);
+      const name = file.substr(0, file.length - ext.length);
       const moduleName = `${packageName}/${name}`;
       if (name === 'index') {
         aliases.push(format(aliasTemplate, moduleName, '', packageName));
@@ -145,17 +181,10 @@ export async function pkgHasFlowFiles(
   projectRoot: string,
   packageName: string,
 ): Promise<boolean> {
-  let pathToPackage = await resolve(packageName, {basedir: projectRoot});
-
-  let dir = path.dirname(pathToPackage);
-  while (path.basename(dir) !== "node_modules") {
-    pathToPackage = dir;
-    dir = path.dirname(pathToPackage);
-
-    if (dir === pathToPackage) {
-      throw new Error("Failed to find package directory");
-    }
-  }
+  let pathToPackage = await resolveToDirPath(
+    packageName,
+    {basedir: projectRoot},
+  );
 
   const files = await glob("**/*.flow", {
     cwd: pathToPackage,
@@ -183,17 +212,10 @@ export async function createStub(
   let pathToPackage = null;
   let version = explicitVersion || null;
   try {
-    pathToPackage = await resolve(packageName, {basedir: process.cwd()});
-
-    let dir = path.dirname(pathToPackage);
-    while (path.basename(dir) !== "node_modules") {
-      pathToPackage = dir;
-      dir = path.dirname(pathToPackage);
-
-      if (dir === pathToPackage) {
-        throw new Error("Failed to find package directory");
-      }
-    }
+    pathToPackage = await resolveToDirPath(
+      packageName,
+      {basedir: process.cwd()},
+    );
 
     files = await glob("**/*.{js,jsx}", {
       cwd: pathToPackage,

